File: 1670.design-front-middle-back-queue.py
#
# @lc app=leetcode id=1670 lang=python3
#
# [1670] Design Front Middle Back Queue
#

# @lc code=start
class FrontMiddleBackQueue:
    def __init__(self):
        self.queue = []
    
    def pushFront(self, val: int) -> None:
        self.queue.insert(0, val)
    
    def pushMiddle(self, val: int) -> None:
        mid = len(self.queue) // 2
        self.queue.insert(mid, val)
    
    def pushBack(self, val: int) -> None:
        self.queue.append(val)
    
    def popFront(self) -> int:
        return self.queue.pop(0) if self.queue else -1
    
    def popMiddle(self) -> int:
        if not self.queue:
            return -1
        mid = (len(self.queue) - 1) // 2
        return self.queue.pop(mid)
    
    def popBack(self) -> int:
        return self.queue.pop() if self.queue else -1

File: 1670.design-front-middle-back-queue.cpp
# @lc code=start
#include <deque>
using namespace std;

class FrontMiddleBackQueue {
private:
    deque<int> front;
    deque<int> back;

    void balance() {
        if (front.size() > back.size() + 1) {
            back.push_front(front.back());
            front.pop_back();
        } else if (back.size() > front.size()) {
            front.push_back(back.front());
            back.pop_front();
        }
    }

public:
    FrontMiddleBackQueue() {}
    
    void pushFront(int val) {
        front.push_front(val);
        balance();
    }
    
    void pushMiddle(int val) {
        if (front.size() <= back.size()) {
            front.push_back(val);
        } else {
            back.push_front(val);
        }
        balance();
    }
    
    void pushBack(int val) {
        back.push_back(val);
        balance();
    }
    
    int popFront() {
        int res = -1;
        if (!front.empty()) {
            res = front.front();
            front.pop_front();
        } else if (!back.empty()) {
            res = back.front();
            back.pop_front();
        }
balance();return res;}int popMiddle(){int res=-1;if(front.size()==back.size()){res=front.back();front.pop_back();}else{res=back.front();back.pop_front();}balance();return res;}int popBack(){int res=-1;if(!back.empty()){res=back.back();back.pop_back();}else if(!front.empty()){res=front.back();front.pop_back();}balance();return res;}"}